pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java

```diff
@@ -63,52 +66,33 @@ public static BlockExchange getExchange(MailboxService<TransferableBlock> mailbo
     }
   }
 
-  protected BlockExchange(MailboxService<TransferableBlock> mailbox, List<MailboxIdentifier> destinations,
-      BlockSplitter splitter) {
-    _mailbox = mailbox;
-    _destinations = destinations;
+  protected BlockExchange(List<SendingMailbox> sendingMailboxes, BlockSplitter splitter) {
+    _sendingMailboxes = sendingMailboxes;
     _splitter = splitter;
   }
 
   public void send(TransferableBlock block) {
     if (block.isEndOfStreamBlock()) {
-      _destinations.forEach(destination -> sendBlock(destination, block));
+      _sendingMailboxes.forEach(destination -> sendBlock(destination, block));
       return;
     }
-
-    Iterator<RoutedBlock> routedBlocks = route(_destinations, block);
-    while (routedBlocks.hasNext()) {
-      RoutedBlock next = routedBlocks.next();
-      sendBlock(next._destination, next._block);
-    }
+    route(_sendingMailboxes, block);
   }
 
-  private void sendBlock(MailboxIdentifier mailboxId, TransferableBlock block) {
-    SendingMailbox<TransferableBlock> sendingMailbox = _mailbox.getSendingMailbox(mailboxId);
-
+  protected void sendBlock(SendingMailbox sendingMailbox, TransferableBlock block) {
     if (block.isEndOfStreamBlock()) {
       sendingMailbox.send(block);
       sendingMailbox.complete();
       return;
     }
 
     DataBlock.Type type = block.getType();
+    // TODO: Move splitter into mailbox service.
```

Comment:
why do we need to move splitter into mailbox service? did you mean mailbox operator?